drivers/mil_passive_sonar/passive_sonar_driver/passive_sonar_driver.py

```diff
@@ -0,0 +1,590 @@
+        1) assign signals to self.signals
+        2) assign tf translation to self.translation
+        3) assign tf rotation to self.rotation
+        4) assign True to self.ready
+        '''
+        pass
+
+    def get_input(self):
+        '''
+        Returns a tuple with 3 things: signals, translation, rotation
+
+        signals - a numpy array with a row for each receiver, and a column for each element of
+            the recorded signal. The first row is reserved for the reference signal. The remaining
+            rows should be in the same order as define in rosparam /passive_sonar/receiver_locations.
+        translation - a 3 element 1D numpy array representing the position of the receiver array
+            in <ref_frame>. The second element should be a
+        rotation - 3x3 rotation matrix (numpy array) representing the orientation of the receiver
+            array in <ref_frame>.
+
+        Derived classes SHOULD NOT OVERRIDE this method.
+        '''
+        return self.signals, self.translation, self.rotation
+
+    def reset(self):
+        '''
+        Driver will reset state when signals are received, could be used as feedback that signals
+        were received.
+        '''
+        self.ready = False
+
+
+class _SerialReceiverArray(ReceiverArrayInterface):
+    '''
+    This is the default serial ReceiverArrayInterface for the passive sonar driver.
+    It is used when the keyword arg input_mode='serial' is passed to the passive sonar
+    driver constructor.
+    '''
+    def __init__(self, param_names, num_receivers):
+        self.num_receivers = num_receivers
+
+        load = lambda prop: setattr(self, prop, rospy.get_param('passive_sonar/' + prop))
+        [load(x) for x in param_names]
+
+        self.tf2_buf = tf2_ros.Buffer()
+
+        try:
+            self.ser = serial.Serial(port=self.port, baudrate=self.baud, timeout=self.read_timeout)
+            self.ser.flushInput()
+        except serial.SerialException, e:
+            rospy.err("Sonar serial connection error: " + str(e))
+            raise e
+
+    def input_request(self):
+        try:
+            self._request_signals()
+            self._receive_signals()
+            T = self.getReceiverPose(rospy.Time.now(), self.receiver_array_frame,
+                                     self.locating_frame)
+            self.translation, self.rotation = T
+            self.ready = True
+        except Exception as e:
+            rospy.logerr(str(e))
+            raise e
+ 
+    def _request_signals(self):
+        '''
+        Request a set of digital signals from a serial port
+
+        serial_port - open instance of serial.Serial
+        data_request_code - char or string to sent to receiver board to signal a data request
+        tx_start_code - char or string that the receiver board will send  us to signal the
+            start of a transmission
+        '''
+        self.ser.flushInput()
+        try:
```

Comment:
Wouldn't you want errors here to propagate up to the try statement in `intput_request` so that it won't do calculations with bad data?